app/controllers/uploads_controller.rb

```diff
@@ -9,48 +9,61 @@ def index
   end
 
   def new
-    csv_type = params[:csv_type]
-
-    @upload = Upload.new(csv_type: csv_type)
-    @upload.skip_lines = defaults(csv_type)['skip_lines']
-
+    @upload = new_upload(params[:csv_type])
     return if @upload.csv_type.blank? || @upload.csv_type_check?
 
-    flash.alert = errors_for_alert([@upload])
+    log_and_display_error(errors_for_alert([@upload]), 'Warning')
     @upload.csv_type = nil
   end
 
   def create
     @upload = Upload.create(merged_params)
-    data = load_csv
 
-    if @upload.ok?
-      failed_instances = data.failed_instances
-      warnings = errors_for_alert(failed_instances)
+    begin
+      data = load_csv
+      display_failed_instances(data.failed_instances)
 
-      Rails.logger.warn warnings
+      redirect_to @upload
+    rescue StandardError => e
+      log_and_display_error(e.message, "Failed to upload #{original_filename}.", e.backtrace)
+      @upload = new_upload(merged_params[:csv_type])
 
-      redirect_to @upload, alert: warnings, notice: message_for_notice(failed_instances)
-    else
-      render :new, alert: errors_for_alert([@upload]), notice: "Failed to upload #{original_filename}."
+      render :new
     end
   end
 
   def show
     @upload = Upload.find_by(id: params[:id])
-    redirect_to uploads_path, alert: ["Upload with id: '#{params[:id]}' not found"] unless @upload
+    return if @upload.present?
```

Comment:
If the upload fails it redirects to the index action, if it succeeds the show view is presented by default.